### DISCUSSION TITLE
package a project with their dependencies

### DISCUSSION THREAD

**Discussion by forty** [pnpm/pnpm#4237] [Q&A] [answered]
Hello,

I'm coming from https://github.com/pnpm/pnpm/issues/2198 where is was recommended to create a discussion :) I read the case, and I'm not sure I found the solution to my issue, so I'm asking here, sorry if the solution was already mentioned, then I misunderstood it.

What I want to do: I want to package one of my package in a tarball (like https://pnpm.io/cli/pack) but also containing all the dependencies. The goal is to have something I can untar on a server and run, without pnpm installed or a populated store directory.

With npm, it was easy, I just tarred both the build files and the node_modules together and it was all good.

With pnpm phantom dependency protection (symlinks), it's not very easy to do. I want to avoid shipping the whole pnpm-store as it might contain useless packages from other project/build.

Am I missing something? Would there be value, for example, in adding an option `--bundle-dependencies` to `pmpm pack`? Or is there already a third party package doing something similar?

I saw the suggestion of 
- using `make-dedicated-lockfile`: I'm not sure how this helps with the bundling?
- using pnp linking: that looks good, but I would still need to somehow gather the right zip files in my bundle?

Thanks!

**Comment by ludofischer** [comment]
It does not directly answer your question, but have you tried using the [pnp linker](https://pnpm.io/npmrc#node-linker)? I think the Yarn maintainers designed that installation strategy exactly for the case where you want to copy the dependency files to production. From reading the docs, it looks like the `hoisted` linker could also help you, since it removes the symlinks and it's recommended if you use `bundledDependencies` in `package.json`.

**Reply by forty** [reply]
Hoisted might work yes, but then I would lose most of pnpm benefits, wouldn't I?

You are not the first to say that PNP  could solve my problem (I mentioned it), but I don't understand how. I understand that the pnp.js file is loading depencies from zip files, but those zip files would still need to be packaged somehow?

**Reply by zkochan** [reply]
If the tarball can pack and unpack the symlinks correctly, you may pack your application with default configuration as well. This won't mean that you pack the whole store. `node_modules/.pnpm` contains only the dependencies of your app.

**Reply by forty** [reply]
oh that would work indeed! thanks a lot, for some reasons I though the links were always pointing to the central store in HOME, but I was completely wrong. Thanks!

**Reply by zkochan** [reply]
They are hard links. They are real files but don't consume additional disk space.

**Reply by forty** [reply]
@zkochan Ok, got it, thanks a lot!

I made some tests, and it looks like that, in a monorepo using workspaces, the node_module/.pnpm will be located in the root package, rather than in all individual packages node_modules. it means that if I tar a single package, I won't get the .pnpm store (and it probably contains too many packages anyway - there are also the dependencies of the other packages of the mono repo)

Is it possible to ask pnpm to create a node_modules/.pnpm/ per package?

**Reply by ludofischer** [reply]
> You are not the first to say that PNP could solve my problem (I mentioned it), but I don't understand how. I understand that the pnp.js file is loading depencies from zip files, but those zip files would still need to be packaged somehow?

I think you can pack the zipfiles themselves inside the tarball.

**Reply by forty** [reply]
@ludofischer I could do that but from my tests I have 2 issues:

- listing the right zip to includes
- like for the virtual store, it looks like in workspaces, .pnp.cjs is only created for the root, rather than for each packages.

**Reply by ludofischer** [reply]
So you want to deploy an application standalone without running a package manager on the deployment environment, but develop the application in a monorepo. I suppose you thought of the following: 
1.copying the application source and package.json in new directory
2. run `npm install` in that new directory
3. tar the contents of the new directory
The problem I see with the above approach is that if the application depends on other packages in the monorepo, the path to these packages might not work when you copy the `package.json` to a new directory.
But what's the advantage for you of developing in a monorepo, if the application is to be deployed separately from the other packages in the monorepo?

**Reply by forty** [reply]
@ludofischer yes exactly. There are several applications in that monorepo and they all depends on some different subset of the packages in that monorepo.

Indeed your suggested approach doesn't work with the local dependencies and also does not use the package lock.

**Reply by ludofischer** [reply]
In a sense what you would like is 'turn off' the monorepo mode when you tar the package for release. The structure that pnpm, yarn 2+ etc. create when in monorepo mode gets in the way at release time. I just say this 'slicing off' a package from a monorepo in a tarball is a bit unusual, but I can see how it would make sense. In a similar scenario, some people bundle all dependencies with a bundler instead of packaging the `node_modules` directory. But that would change your deployment artifact, which would probably require some testing to ensure nothing breaks in the bundling process.

**Reply by forty** [reply]
Exactly, in the frontend world people seem to be using bundler (ie webpack) to solve this issue. I'm trying to avoid it if possible because it adds it's own complexity, but if not I guess I'll just end up biting the bullet.

**Reply by ludofischer** [reply]
Some people do bundle their server-side code. For example, there's https://github.com/vercel/ncc that is (was?) used to bundle the Next.js framework. Since presumably you don't need to transpile anything, bundling might not be too hard to configure.

**Reply by forty** [reply]
After looking a bit more, I went ahead, and starting looking into how I would do that if I had to code it.

I got this https://github.com/forty/pnpm-bundler (single file mess, sorry https://github.com/forty/pnpm-bundler/blob/master/index.js).

In theory that would allow to build a dependency-free pnpm tarball for exemple, without esbuild which is currently used (maybe it's not a good use case and esbuild works well here, but I used the pnpm repo as a test repository ^^)

@zkochan don't you think this kind of feature would be interesting? At least I think it would be useful for us, but maybe we are doing weird things ^^
Anyway, I'll probably continue improving that script if not, in case someone else find it useful.

**Reply by forty** [reply]
(and @ludofischer I'm not really fan of real bundling - at least in theory - because of the mess it can cause with stack trace, then you have to work with source map, and then the code you have in prod is not the same that the one you tested, etc etc. I guess it's all solved problems, but I'd rather avoid it if I can)

**Reply by zkochan** [reply]
I use this lib to pack pnpm: https://github.com/zkochan/publish-packed

**Reply by forty** [reply]
Thanks for your answers @zkochan !

I have been looking into it a bit more to understand what's going on.

I see that the package `packages/pnpm` is using `publish-packed` indeed, but is also bundled with esbuild. Is it only used for node-gyp because this is native stuff that cannot be bundled by esbuild?

I think that use case wit a single dependency might work well, however isn't there a risk that the dependency tree in the package ends up being very different because publish-packed forces --hoisted ? Also I think all the features that make pnpm better (no phantom deps, no doppleganger, etc etc) still make sense in the packaging

**Comment by minlare** [comment, marked as answer]
[`pnpm deploy`](https://pnpm.io/cli/deploy) is the new way to do this - https://github.com/pnpm/pnpm/issues/4378#issuecomment-1167758599

**Comment by XingXiaoWu** [comment]
same question

**Reply by forty** [reply]
As mentioned on the comment above, pnpm deploy now does this.

**Reply by XingXiaoWu** [reply]
I'm not workspace,can't use deploy
